File: bkbook/book.py
# SPDX-License-Identifier: BSD-3-Clause
# Copyright (c) 2026 satorux
# Copyright (c) 1997-2006 Motoyuki Kasahara

"""本——``catalog`` / ``catalogs`` ファイルと、そこに並ぶサブブック。

**本**（book）はディレクトリツリー 1 つ、たいていは CD-ROM 1 枚にあたる。
その中に **サブブック** が 1 つ以上あり、それぞれが独自の索引と本文を持つ
独立した辞書である。ディスク上の配置は 2 通り:

``EB``（電子ブック）
    直下に ``catalog``、サブブックのデータは ``<dir>/start``。
``EPWING``
    直下に ``catalogs``、サブブックのデータは ``<dir>/data/honmon``、
    フォントは ``<dir>/gaiji/``。
"""

from __future__ import annotations

import os

from .jacode import (
    CHARCODE_ISO8859_1,
    CHARCODE_JISX0208,
    CHARCODE_JISX0208_GB2312,
    CHARCODE_NAMES,
    decode_title,
)
from .subbook import SubBook
from .zio import EbError, Zio, find_file

DISC_EB = "eb"
DISC_EPWING = "epwing"

CATALOG_HEADER_SIZE = 16
EB_CATALOG_ENTRY_SIZE = 40
EPWING_CATALOG_ENTRY_SIZE = 164

EB_TITLE_LENGTH = 30
EPWING_TITLE_LENGTH = 80
DIRECTORY_NAME_LENGTH = 8

MAX_SUBBOOKS = 50
MAX_FONTS = 4


class BookError(EbError):
    """本を開けないか、catalog が壊れている。"""


def _trimmed_name(data: bytes) -> str:
    """固定長 ASCII の名前欄を復号する。NUL か空白で切る。"""
    text = data.decode("ascii", errors="replace")
    for terminator in ("\0", " "):
        index = text.find(terminator)
        if index >= 0:
            text = text[:index]
    return text


def is_book(path: str) -> bool:
    """``path`` に本があるか——つまり catalog があるか。"""
    return os.path.isdir(path) and bool(
        find_file(path, "catalog") or find_file(path, "catalogs")
    )


def find_books(path: str) -> list[str]:
    """``path`` にある本、または ``path`` の直下にある本すべて。

    コレクションをまとめて指定できるようにするためのもの。ディスクをコピーした
    ディレクトリを指せば、1 冊ずつ並べなくても全部が返る。
    """
    if is_book(path):
        return [path]
    if not os.path.isdir(path):
        return []
    return [
        entry_path
        for entry in sorted(os.listdir(path))
        for entry_path in [os.path.join(path, entry)]
        if is_book(entry_path)
    ]


class Book:
    """開かれた EB または EPWING の本。"""

    def __init__(self, path: str):
        self.path = os.path.abspath(path)
        if not os.path.isdir(self.path):
            raise BookError(f"{path}: not a directory")

        self.character_code = self._load_language()
        self.disc_code, catalog_path = self._find_catalog()
        self.epwing_version = 0
        self.subbooks: list[SubBook] = []
        self._load_catalog(catalog_path)

    # -- catalog ---------------------------------------------------------

    def _find_catalog(self) -> tuple[str, str]:
        catalog = find_file(self.path, "catalog")
        if catalog is not None:
            return DISC_EB, catalog
        catalog = find_file(self.path, "catalogs")
        if catalog is not None:
            return DISC_EPWING, catalog
        raise BookError(f"{self.path}: no catalog or catalogs file")

    def _load_language(self) -> int:
        """``language`` からその本の文字コードを読む。

        このファイルは EB のみ、しかも任意。規格上の既定は JIS X 0208 で、
        libeb もそう仮定している。
        """
        path = find_file(self.path, "language")
        if path is None:
            return CHARCODE_JISX0208
        try:
            with Zio(path) as zio:
                header = zio.read(0, CATALOG_HEADER_SIZE)
        except EbError:
            return CHARCODE_JISX0208
        if len(header) < 2:
            return CHARCODE_JISX0208
        code = int.from_bytes(header[0:2], "big")
        if code not in (
            CHARCODE_ISO8859_1,
            CHARCODE_JISX0208,
            CHARCODE_JISX0208_GB2312,
        ):
            return CHARCODE_JISX0208
        return code

    def _load_catalog(self, catalog_path: str) -> None:
        is_eb = self.disc_code == DISC_EB
        entry_size = EB_CATALOG_ENTRY_SIZE if is_eb else EPWING_CATALOG_ENTRY_SIZE
        title_length = EB_TITLE_LENGTH if is_eb else EPWING_TITLE_LENGTH

        with Zio(catalog_path) as zio:
            header = zio.read(0, CATALOG_HEADER_SIZE)
            if len(header) < CATALOG_HEADER_SIZE:
                raise BookError(f"{catalog_path}: truncated catalog")

            count = int.from_bytes(header[0:2], "big")
            if count == 0:
                raise BookError(f"{catalog_path}: catalog lists no subbooks")
            count = min(count, MAX_SUBBOOKS)

            if not is_eb:
                self.epwing_version = int.from_bytes(header[2:4], "big")

            for index in range(count):
                offset = CATALOG_HEADER_SIZE + index * entry_size
                entry = zio.read(offset, entry_size)
                if len(entry) < entry_size:
                    raise BookError(
                        f"{catalog_path}: truncated entry for subbook {index}"
                    )
                self.subbooks.append(self._parse_entry(index, entry, title_length))

    def _parse_entry(self, index: int, entry: bytes, title_length: int) -> SubBook:
        title = decode_title(entry[2 : 2 + title_length], self.character_code)
        name_at = 2 + title_length
        directory_name = _trimmed_name(
            entry[name_at : name_at + DIRECTORY_NAME_LENGTH]
        )

        if self.disc_code == DISC_EB:
            index_page = 1
            narrow_fonts: list[str] = []
            wide_fonts: list[str] = []
        else:
            index_page = int.from_bytes(
                entry[name_at + DIRECTORY_NAME_LENGTH + 4 :][:2], "big"
            )
            # フォントのファイル名はタイトルから固定の位置にある。全角が
            # +18、半角が +50、それぞれ 8 バイトの枠が 4 つずつ。
            wide_fonts = _font_names(entry, 2 + title_length + 18)
            narrow_fonts = _font_names(entry, 2 + title_length + 50)

        return SubBook(
            book=self,
            code=index,
            title=title,
            directory_name=directory_name,
            index_page=index_page,
            narrow_font_names=narrow_fonts,
            wide_font_names=wide_fonts,
        )

    # -- convenience -----------------------------------------------------

    @property
    def character_code_name(self) -> str:
        return CHARCODE_NAMES.get(self.character_code, "unknown")

    def subbook(self, which: int | str) -> SubBook:
        """サブブックを番号、ディレクトリ名、または完全なタイトルで引く。"""
        if isinstance(which, int):
            return self.subbooks[which]
        lowered = which.lower()
        for subbook in self.subbooks:
            if subbook.directory_name.lower() == lowered or subbook.title == which:
                return subbook
        raise KeyError(which)

    def close(self) -> None:
        for subbook in self.subbooks:
            subbook.close()

    def __enter__(self) -> "Book":
        return self

    def __exit__(self, *exc) -> None:
        self.close()

    def __repr__(self) -> str:
        return (
            f"<Book {self.path!r} {self.disc_code} "
            f"{self.character_code_name} subbooks={len(self.subbooks)}>"
        )


def _font_names(entry: bytes, offset: int) -> list[str]:
    """8 バイトのフォント名の枠を 4 つ読む。空の枠は飛ばす。"""
    names = []
    for index in range(MAX_FONTS):
        at = offset + index * DIRECTORY_NAME_LENGTH
        slot = entry[at : at + DIRECTORY_NAME_LENGTH]
        if not slot or slot[0] == 0 or slot[0] >= 0x80:
            names.append("")
            continue
        names.append(_trimmed_name(slot))
    return names

File: web/bkbook/book.js
// SPDX-License-Identifier: BSD-3-Clause
// Copyright (c) 2026 satorux
// Copyright (c) 1997-2006 Motoyuki Kasahara

/**
 * 本——catalog / catalogs ファイルと、そこに並ぶサブブック。Python 版 book.py の移植。
 */

import {
  CHARCODE_ISO8859_1,
  CHARCODE_JISX0208,
  CHARCODE_JISX0208_GB2312,
  CHARCODE_NAMES,
  decodeTitle,
} from "./jacode.js";
import { SubBook } from "./subbook.js";
import { EbError, Zio, be, findFile } from "./zio.js";

export const DISC_EB = "eb";
export const DISC_EPWING = "epwing";

const CATALOG_HEADER_SIZE = 16;
const EB_CATALOG_ENTRY_SIZE = 40;
const EPWING_CATALOG_ENTRY_SIZE = 164;

const EB_TITLE_LENGTH = 30;
const EPWING_TITLE_LENGTH = 80;
const DIRECTORY_NAME_LENGTH = 8;

const MAX_SUBBOOKS = 50;
const MAX_FONTS = 4;

/** 本を開けないか、catalog が壊れている。 */
export class BookError extends EbError {}

/** 固定長 ASCII の名前欄を復号する。NUL か空白で切る。 */
function trimmedName(data) {
  let text = "";
  for (let i = 0; i < data.length; i++) text += String.fromCharCode(data[i]);
  for (const terminator of ["\0", " "]) {
    const index = text.indexOf(terminator);
    if (index >= 0) text = text.slice(0, index);
  }
  return text;
}

/** path に本があるか——つまり catalog があるか。 */
export async function isBook(fs, path) {
  if (!(await fs.isdir(path))) return false;
  return Boolean((await findFile(fs, path, "catalog")) || (await findFile(fs, path, "catalogs")));
}

/** path にある本、または path の直下にある本すべて。 */
export async function findBooks(fs, path) {
  if (await isBook(fs, path)) return [path];
  if (!(await fs.isdir(path))) return [];
  const entries = (await fs.listdir(path)) || [];
  entries.sort();
  const books = [];
  for (const entry of entries) {
    const entryPath = fs.join(path, entry);
    if (await isBook(fs, entryPath)) books.push(entryPath);
  }
  return books;
}

/** 開かれた EB または EPWING の本。 */
export class Book {
  constructor(fs, path) {
    this.fs = fs;
    this.path = path;
    this.characterCode = CHARCODE_JISX0208;
    this.discCode = DISC_EB;
    this.epwingVersion = 0;
    this.subbooks = [];
  }

  static async open(fs, path) {
    if (!(await fs.isdir(path))) throw new BookError(`${path}: not a directory`);
    const book = new Book(fs, path);
    book.characterCode = await book._loadLanguage();
    const [discCode, catalogPath] = await book._findCatalog();
    book.discCode = discCode;
    await book._loadCatalog(catalogPath);
    return book;
  }

  async _findCatalog() {
    let catalog = await findFile(this.fs, this.path, "catalog");
    if (catalog !== null) return [DISC_EB, catalog];
    catalog = await findFile(this.fs, this.path, "catalogs");
    if (catalog !== null) return [DISC_EPWING, catalog];
    throw new BookError(`${this.path}: no catalog or catalogs file`);
  }

  async _loadLanguage() {
    const path = await findFile(this.fs, this.path, "language");
    if (path === null) return CHARCODE_JISX0208;
    let header;
    try {
      const zio = await Zio.open(this.fs, path);
      try {
        header = await zio.read(0, CATALOG_HEADER_SIZE);
      } finally {
        await zio.close();
      }
    } catch (error) {
      if (error instanceof EbError) return CHARCODE_JISX0208;
      throw error;
    }
    if (header.length < 2) return CHARCODE_JISX0208;
    const code = be(header, 0, 2);
    if (![CHARCODE_ISO8859_1, CHARCODE_JISX0208, CHARCODE_JISX0208_GB2312].includes(code)) {
      return CHARCODE_JISX0208;
    }
    return code;
  }

  async _loadCatalog(catalogPath) {
    const isEb = this.discCode === DISC_EB;
    const entrySize = isEb ? EB_CATALOG_ENTRY_SIZE : EPWING_CATALOG_ENTRY_SIZE;
    const titleLength = isEb ? EB_TITLE_LENGTH : EPWING_TITLE_LENGTH;

    const zio = await Zio.open(this.fs, catalogPath);
    try {
      const header = await zio.read(0, CATALOG_HEADER_SIZE);
      if (header.length < CATALOG_HEADER_SIZE) {
        throw new BookError(`${catalogPath}: truncated catalog`);
      }
      let count = be(header, 0, 2);
      if (count === 0) throw new BookError(`${catalogPath}: catalog lists no subbooks`);
      count = Math.min(count, MAX_SUBBOOKS);
      if (!isEb) this.epwingVersion = be(header, 2, 2);

      for (let index = 0; index < count; index++) {
        const offset = CATALOG_HEADER_SIZE + index * entrySize;
        const entry = await zio.read(offset, entrySize);
        if (entry.length < entrySize) {
          throw new BookError(`${catalogPath}: truncated entry for subbook ${index}`);
        }
        this.subbooks.push(this._parseEntry(index, entry, titleLength));
      }
    } finally {
      await zio.close();
    }
  }

  _parseEntry(index, entry, titleLength) {
    const title = decodeTitle(entry.subarray(2, 2 + titleLength), this.characterCode);
    const nameAt = 2 + titleLength;
    const directoryName = trimmedName(entry.subarray(nameAt, nameAt + DIRECTORY_NAME_LENGTH));

    let indexPage = 1;
    let narrowFonts = [];
    let wideFonts = [];
    if (this.discCode !== DISC_EB) {
      indexPage = be(entry, nameAt + DIRECTORY_NAME_LENGTH + 4, 2);
      wideFonts = fontNames(entry, 2 + titleLength + 18);
      narrowFonts = fontNames(entry, 2 + titleLength + 50);
    }
    return new SubBook(this, {
      code: index,
      title,
      directoryName,
      indexPage,
      narrowFontNames: narrowFonts,
      wideFontNames: wideFonts,
    });
  }

  get characterCodeName() {
    return CHARCODE_NAMES[this.characterCode] || "unknown";
  }

  /** サブブックを番号、ディレクトリ名、または完全なタイトルで引く。 */
  subbook(which) {
    if (typeof which === "number") return this.subbooks[which];
    const lowered = which.toLowerCase();
    for (const subbook of this.subbooks) {
      if (subbook.directoryName.toLowerCase() === lowered || subbook.title === which) return subbook;
    }
    throw new Error(`no subbook ${which}`);
  }

  async close() {
    for (const subbook of this.subbooks) await subbook.close();
  }
}

/** 8 バイトのフォント名の枠を 4 つ読む。空の枠は "" にする。 */
function fontNames(entry, offset) {
  const names = [];
  for (let index = 0; index < MAX_FONTS; index++) {
    const at = offset + index * DIRECTORY_NAME_LENGTH;
    const slot = entry.subarray(at, at + DIRECTORY_NAME_LENGTH);
    if (slot.length === 0 || slot[0] === 0 || slot[0] >= 0x80) {
      names.push("");
      continue;
    }
    names.push(trimmedName(slot));
  }
  return names;
}
